### PR TITLE
[spirv] Translate operator[] for vector and matrix types

### PR DESCRIPTION
**Pull request by antiagainst**

### CODE REVIEW

File: tools/clang/lib/SPIRV/TypeTranslator.cpp [hunk 1]
```diff
@@ -14,6 +14,13 @@ namespace clang {
 namespace spirv {
 
 uint32_t TypeTranslator::translateType(QualType type) {
+  {
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/TypeTranslator.cpp]: I'm not a fan of the extra braces here. I think the function would be easier to read without them.

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp [hunk 2]
```diff
@@ -2063,7 +2063,138 @@ class SPIRVEmitter : public ASTConsumer {
...
+  /// matrix using operator[].
+  /// On success, writes the base vector/matrix into *base, and the indices into
+  /// *index0 and *index1, if there are two levels of indexing. If there is only
+  /// one level of indexing, writes the index into *index0 and nullptr into
+  /// *index1.
+  /// Assumes base, index0, and index1 are not nullptr.
+  bool isVecMatIndexing(const CXXOperatorCallExpr *vecIndexExpr,
+                        const Expr **base, const Expr **index0,
+                        const Expr **index1) {
+    // matrix [] []         vector []
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: The comment here is a bit unclear and can be improved. e.g.

Is it `matrix[index0][index1]`, or `matrix[index1][index0]`

IIUC:
`vector = matrix[index0]`
and
`scalar = vector[index1]`

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp [hunk 3]
```diff
@@ -2063,7 +2063,138 @@ class SPIRVEmitter : public ASTConsumer {
...
+            indices.push_back(doExpr(index0Expr));
+          }
+        }
+
+        if (indices.size() == 0) {
+          return doExpr(baseExpr);
+        }
+
+        uint32_t base = doExpr(baseExpr);
+        // If we are indexing into a rvalue, to use OpAccessChain, we need first
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: s/we need first to/we first need to/

File: tools/clang/test/CodeGenSPIRV/op.vector.access.hlsl [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+// CHECK-NEXT: [[b1:%\d+]] = OpLoad %float %b
+// CHECK-NEXT: OpStore %s [[b1]]
+    s = b[index];
+// CHECK-NEXT: [[s3:%\d+]] = OpLoad %float %s
+// CHECK-NEXT: OpStore %b [[s3]]
+    b[index] = s;
+
+    // From rvalue
+    //s = (a + a)[0];
```
Comment (ehsannas) [inline tools/clang/test/CodeGenSPIRV/op.vector.access.hlsl]: no SPIR-V for these?

File: tools/clang/lib/SPIRV/TypeTranslator.cpp (hunk 1 above)
Comment (antiagainst) [inline tools/clang/lib/SPIRV/TypeTranslator.cpp]: It was there to give `canonicalType` a much smaller scope. But I'm fine to remove it. Will do.

Comment (antiagainst) [inline tools/clang/lib/SPIRV/TypeTranslator.cpp]: Done

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp (hunk 2 above)
Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: Good call! Done.

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp (hunk 3 above)
Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: Yep. Done.

File: tools/clang/test/CodeGenSPIRV/op.vector.access.hlsl (hunk 4 above)
Comment (antiagainst) [inline tools/clang/test/CodeGenSPIRV/op.vector.access.hlsl]: Good catch. Somehow forgot this. Added.

**Pull request merged**